com.unity.netcode.gameobjects/Runtime/SceneManagement/NetworkSceneManager.cs

```diff
@@ -219,261 +552,876 @@ internal void OnSceneSwitch(uint sceneIndex, Guid switchSceneGuid, Stream object
+            }
+        }
+
+        /// <summary>
+        /// Both Client and Server: Incoming scene event entry point
+        /// </summary>
+        /// <param name="clientId">client who sent the scene event</param>
+        /// <param name="stream">data associated with the scene event</param>
+        internal void HandleSceneEvent(ulong clientId, Stream stream)
         {
-            foreach (var switchSceneProgress in SceneSwitchProgresses.Values)
+            if (m_NetworkManager != null)
+            {
+                if (stream != null)
+                {
+                    var reader = NetworkReaderPool.GetReader(stream);
+                    SceneEventData.OnRead(reader);
+                    NetworkReaderPool.PutBackInPool(reader);
+                    if (SceneEventData.IsSceneEventClientSide())
+                    {
+                        HandleClientSceneEvent(stream);
+                    }
+                    else
+                    {
+                        HandleServerSceneEvent(clientId, stream);
+                    }
+                }
+                else
+                {
+                    Debug.LogError($"Scene Event {nameof(OnClientSceneLoadingEvent)} was invoked with a null stream!");
+                    return;
+                }
+            }
+            else
             {
-                switchSceneProgress.RemoveClientAsDone(clientId);
+                Debug.LogError($"{nameof(NetworkSceneManager.HandleSceneEvent)} was invoked but {nameof(NetworkManager)} reference was null!");
             }
         }
 
+        /// <summary>
+        /// Moves all NetworkObjects that don't have the <see cref="NetworkObject.DestroyWithScene"/> set to
+        /// the "Do not destroy on load" scene.
+        /// </summary>
         private void MoveObjectsToDontDestroyOnLoad()
         {
             // Move ALL NetworkObjects to the temp scene
-            var objectsToKeep = m_NetworkManager.SpawnManager.SpawnedObjectsList;
+            var objectsToKeep = new HashSet<NetworkObject>(m_NetworkManager.SpawnManager.SpawnedObjectsList);
 
             foreach (var sobj in objectsToKeep)
             {
-                //In case an object has been set as a child of another object it has to be unchilded in order to be moved from one scene to another.
+                //In case an object has been set as a child of another object it has to be removed from the parent in order to be moved from one scene to another.
                 if (sobj.gameObject.transform.parent != null)
                 {
                     sobj.gameObject.transform.parent = null;
                 }
 
-                UnityEngine.Object.DontDestroyOnLoad(sobj.gameObject);
+                if (!sobj.DestroyWithScene)
+                {
+                    UnityEngine.Object.DontDestroyOnLoad(sobj.gameObject);
+                }
+                else if (m_NetworkManager.IsServer)
+                {
+                    sobj.Despawn(true);
```

Comment:
I don't know enough about how network objects should behave in this scenario, but my understanding of this function is that it should be stopping network objects from despawning by moving them into DontDestroyOnLoad, so it's odd that we then are despawning stuff in here.